### PR TITLE
TF-3275 Fix FCM iOS foreground desync

### PR DESCRIPTION
**Pull request by tddang-linagora**
## Issue
- #3307 

## Root cause
- Notification from FCM contains `mutable-content: true`
- Those notification will be transfered to Notification Service Extension ([See here](https://developer.apple.com/documentation/usernotifications/modifying-content-in-newly-delivered-notifications#Configure-the-payload-for-the-remote-notification)) whether the app is in background or foreground
- Due to the app is in foreground, no notification will be shown
- Due to NSE handle the data, FCM on Flutter side cannot handle it

## Resolved

https://github.com/user-attachments/assets/ec20ffec-f9bb-4e58-a8e2-02a6db4adcf0

### CODE REVIEW

File: ios/Runner/AppDelegate.swift [hunk 1]
```diff
@@ -97,6 +97,16 @@ import flutter_local_notifications
...
         if validateDisplayPushNotification(userInfo: notification.request.content.userInfo) {
             completionHandler([.alert, .badge, .sound])
+        } else if UIApplication.shared.applicationState == .active {
+            let viewController = window?.rootViewController as! FlutterViewController
+            let fcmMethodChannel = FlutterMethodChannel(
+                name: CoreUtils.FCM_METHOD_CHANNEL_NAME,
+                binaryMessenger: viewController.binaryMessenger)
+            fcmMethodChannel.invokeMethod(
+                CoreUtils.FCM_ON_MESSAGE_METHOD_NAME,
+                arguments: notification.request.content.userInfo)
```
Comment (dab246) [inline ios/Runner/AppDelegate.swift:107]: We should initial `MethodChannel` in `didFinishLaunchingWithOptions `

File: ios/TwakeMailNSE/NotificationService.swift [hunk 2]
```diff
@@ -143,6 +145,7 @@ class NotificationService: UNNotificationServiceExtension {
         self.modifiedContent?.body = message
         self.modifiedContent?.badge = NSNumber(value: 1)
         self.modifiedContent?.sound = .default
+        self.modifiedContent?.userInfo = ["data": originalContent.userInfo]
```
Comment (dab246) [inline ios/TwakeMailNSE/NotificationService.swift:148]: why do we need `data` key?

File: ios/TwakeMailNSE/NotificationService.swift [hunk 3]
```diff
@@ -155,7 +158,9 @@ class NotificationService: UNNotificationServiceExtension {
         self.modifiedContent?.body = body ?? ""
         self.modifiedContent?.badge = NSNumber(value: badgeCount)
         self.modifiedContent?.sound = .default
-        self.modifiedContent?.userInfo = userInfo
+        var combinedUserInfo = userInfo
+        combinedUserInfo["data"] = originalContent.userInfo
+        self.modifiedContent?.userInfo = combinedUserInfo
```
Comment (dab246) [inline ios/TwakeMailNSE/NotificationService.swift:163]: Also please test notification on onBackground

File: ios/TwakeMailNSE/NotificationService.swift (hunk 2 above)
Comment (tddang-linagora) [reply ios/TwakeMailNSE/NotificationService.swift:148]: `onMessage` of FCM Flutter is using data key for its payload.

File: ios/TwakeMailNSE/NotificationService.swift (hunk 3 above)
Comment (tddang-linagora) [reply ios/TwakeMailNSE/NotificationService.swift:163]: FCM background is not affected

File: ios/Runner/AppDelegate.swift (hunk 1 above)
Comment (tddang-linagora) [reply ios/Runner/AppDelegate.swift]: Done

File: ios/TwakeMailNSE/NotificationService.swift (hunk 3 above)
Comment (dab246) [inline ios/TwakeMailNSE/NotificationService.swift:163]:
```suggestion
modifiedContent.userInfo = userInfo.merging(["data": originalContent.userInfo]) { (_, new) in new }
```

Comment (tddang-linagora) [reply ios/TwakeMailNSE/NotificationService.swift]: Done

**Comment by dab246** [comment]
- [ ] Tested on `tmail.linagora.com` environment. Real device `iOS 18.1.1`. It still doesn't work.

Screenshot 2024-12-02 at 15 07 26

IMG_9764

**Comment by tddang-linagora** [comment]
>It still doesn't work.

https://github.com/user-attachments/assets/7ce8a063-9ec7-4343-b6fc-a48cf873d709

`tmail.linagora.com` worked fine

**Comment by hoangdat** [comment]
did you tested with iOS v18.* @tddang-linagora

**Comment by tddang-linagora** [comment]
> did you tested with iOS v18.* @tddang-linagora

I don't have a iOS 18 device. Can this be tested on simulator?

File: ios/Runner/AppDelegate.swift [hunk 4]
```diff
@@ -97,6 +99,12 @@ import flutter_local_notifications
         }
         if validateDisplayPushNotification(userInfo: notification.request.content.userInfo) {
             completionHandler([.alert, .badge, .sound])
+        } else if UIApplication.shared.applicationState == .active {
```
Comment (dab246) [inline ios/Runner/AppDelegate.swift:102]: please check it first

**Review by dab246 — COMMENTED** [review]
- [ ] write ADR for this case

File: ios/Runner/AppDelegate.swift (hunk 4 above)
Comment (tddang-linagora) [reply ios/Runner/AppDelegate.swift]: Done

File: ios/TwakeMailNSE/NotificationService.swift [hunk 5]
```diff
@@ -16,6 +16,15 @@ class NotificationService: UNNotificationServiceExtension {
         handler = contentHandler
         modifiedContent = (request.content.mutableCopy() as? UNMutableNotificationContent)
         
+        let appGroupId = "group.com.linagora.teammail"
```
[the remark this answers is not in the document]
Comment (tddang-linagora) [reply ios/TwakeMailNSE/NotificationService.swift]: Done

File: ios/TwakeMailNSE/NotificationService.swift [hunk 6]
```diff
@@ -16,6 +16,15 @@ class NotificationService: UNNotificationServiceExtension {
...
+        let userDefaults = UserDefaults(suiteName: appGroupId)
+        let isAppActive = userDefaults?.value(forKey: CoreUtils.APPLICATION_STATE) as? Bool
+        if let appIsActive = isAppActive, appIsActive {
```
[the remark this answers is not in the document]
Comment (tddang-linagora) [reply ios/TwakeMailNSE/NotificationService.swift]: Done

File: ios/TwakeMailNSE/NotificationService.swift [hunk 7]
```diff
@@ -16,6 +16,15 @@ class NotificationService: UNNotificationServiceExtension {
...
+            var originalContent = request.content.mutableCopy() as? UNMutableNotificationContent
```
[the remark this answers is not in the document]
Comment (tddang-linagora) [reply ios/TwakeMailNSE/NotificationService.swift]: Done

**Review by dab246 — DISMISSED** [review]
- [ ] Add ADR

File: docs/adr/0055-ios-fcm-routing.md [hunk 8]
```diff
@@ -0,0 +1,24 @@
+# 54. Update Flutter Service Worker Initialization
```
Comment (hoangdat) [inline docs/adr/0055-ios-fcm-routing.md:1]:
```suggestion
# 55. iOS FCM Routing
```

**Review by hoangdat — APPROVED** [review]

**Pull request merged**